Allow X509Chain to replace the root trust list #16364

zwread opened the pull request:
Fixes: #16364

zachariahread committed:
Add api interface and windows/linux/osx impl

Add trustMode and customTrustStore to api interface.
Modify implementations for updated interfaces.
Add tests for invalid configurations and to test new
implementations.

zachariahread committed:
Add test for ensuring trust persists for only a single call

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=40379) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=40379) before we can accept your contribution.<br/><br/>:x: zwread [sign now](https://cla.dotnetfoundation.org/dotnet/corefx?pullRequest=40379)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/corefx?pullRequest=40379) it.</sub>

zachariahread committed:
Remove Exclusive Flag From Windows Impl

stephentoub commented on src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_x509chain.c:
```diff
@@ -264,3 +264,17 @@ int32_t AppleCryptoNative_GetOSStatusForChainStatus(PAL_X509ChainStatusFlags cha
+    if (anchorCertificates == NULL)
+    {
+        return;
+    }
```
Are null chains and anchorCertificates considered valid?  If not, should we be returning an error of some kind?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -51,6 +53,12 @@ internal SecTrustChainPal()
+                SafeCreateHandle customCertsArray = null;
+                customCertsArray = PrepareCustomCertsArray(customTrustStore);
+                Interop.AppleCrypto.X509ChainSetTrustAnchorCertificates(chain, customCertsArray);
+            }
```
Nit: please add a blank line after the closing brace

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -51,6 +53,12 @@ internal SecTrustChainPal()
+            if (trustMode == X509ChainTrustMode.CustomRootTrust)
+            {
+                SafeCreateHandle customCertsArray = null;
+                customCertsArray = PrepareCustomCertsArray(customTrustStore);
```
```C#
SafeCreateHandle customCertsArray = PrepareCustomCertsArray(customTrustStore);
```

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -51,6 +53,12 @@ internal SecTrustChainPal()
+            {
+                SafeCreateHandle customCertsArray = null;
+                customCertsArray = PrepareCustomCertsArray(customTrustStore);
+                Interop.AppleCrypto.X509ChainSetTrustAnchorCertificates(chain, customCertsArray);
```
Who owns the SafeCreateHandle at this point?  Was ownership transferred?  Should we be disposing of it?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
-            SafeHandle[] safeHandles = new SafeHandle[ptrs.Length];
-
+            int certificateCount = 1 + (extraStore?.Count ?? 0) + (trustMode == X509ChainTrustMode.CustomRootTrust ? customTrustStore?.Count ?? 0 : 0);
+            int certificatesAdded = 1;
```
Please move this down to just after the `safeHandles[0] = applePal.CertificateHandle;` line, so that it's clear the added count is being set to 1 because we just stored an element into the array.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
-                    AppleCertificatePal extraCertPal = (AppleCertificatePal)extraStore[i].Pal;
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)extraCert.Pal;
+                    safeHandles[certificatesAdded] = extraCertPal.CertificateHandle;
+                    certificatesAdded++;
```
This can be:
```C#
safeHandles[certificatesAdded++] = extraCertPal.CertificateHandle;
```

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
+                for (int i = 0; i < customTrustStore.Count; i++)
+                {
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)customTrustStore[i].Pal;
+                    safeHandles[certificatesAdded + i] = extraCertPal.CertificateHandle;
```
Instead of `certificatesAdded + i`, how about instead doing `certificatesAdded++`, and then down below before we return, doing `Debug.Assert(certificatesAdded == certificateCount);`?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
+        private SafeCreateHandle PrepareCustomCertsArray(X509Certificate2Collection customTrustStore)
+        {
+            X509Certificate2Collection rootCertificates = new X509Certificate2Collection();
+            SafeHandle[] safeHandles;
```
This can be declared when it's assigned below.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
+            return GetCertsArray(safeHandles);
+        }
+
+        private SafeCreateHandle GetCertsArray(SafeHandle[] safeHandles)
```
The implementation of this method makes me a little nervous, and has some duplicated code.  It's pre-existing your PR, but either here or subsequently it'd be nice to see it cleaned up.  It could be written more like (haven't compiled or tested this):
```C#
        private SafeCreateHandle GetCertsArray(SafeHandle[] safeHandles)
        {
            int idx = 0;
            try
            {
                var ptrs = new IntPtr[safeHandles.Length];
                for (; idx < safeHandles.Length; idx++)
                {
                    SafeHandle handle = safeHandles[idx];
                    bool addedRef = false;
                    handle.DangerousAddRef(ref addedRef);
                    ptrs[idx] = handle.DangerousGetHandle();
                }

                // Creating the array has the effect of calling CFRetain() on all of the pointers, so the native
                // resource is safe even when we DangerousRelease=>ReleaseHandle them.
                SafeCreateHandle certsArray = Interop.CoreFoundation.CFArrayCreate(ptrs, (UIntPtr)ptrs.Length);
                _extraHandles.Push(certsArray);
                return certsArray;
            }
            finally
            {
                for (idx--; idx >= 0; idx--)
                {
                    safeHandles[idx].DangerousRelease();
                }
            }
        }
```

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:
```diff
@@ -125,6 +128,36 @@ public SafeX509ChainHandle SafeHandle
+                    else
+                    {
+                        AddToStackAndUpRef(((OpenSslX509CertificateReader)cert.Pal).SafeHandle, untrusted);
+                    }
```
This can be simplified to:
```C#
AddToStackAndUpRef(((OpenSslX509CertificateReader)cert.Pal).SafeHandle, 
    cert.SubjectName.RawData.ContentsEqual(cert.IssuerName.RawData) ? customTrust : untrusted);
```
or if you prefer less in a single statement:
```C#
SafeX509StackHandle toAdd = cert.SubjectName.RawData.ContentsEqual(cert.IssuerName.RawData) ? customTrust : untrusted;
AddToStackAndUpRef(((OpenSslX509CertificateReader)cert.Pal).SafeHandle, toAdd);
```
Either way, we avoid the duplicated calls to AddToStackAndUpRef.

bartonjs replied:
(Personally, I prefer the latter (saving the `toAdd` local))

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:
```diff
@@ -125,6 +128,36 @@ public SafeX509ChainHandle SafeHandle
+                Interop.Crypto.X509StackAddMultiple(untrusted, systemTrust);
+                Interop.Crypto.X509StackAddMultiple(untrusted, s_userRootStore.GetNativeCollection());
+
+                return Interop.Crypto.X509ChainNew(customTrust, Interop.Crypto.NewX509Stack());
```
What happens to `customTrust` after this point?  Do we need to dispose of it?  Is ownership transferred?

stephentoub replied:
Same question for the result of XewX509Stack().

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:
```diff
@@ -125,6 +128,36 @@ public SafeX509ChainHandle SafeHandle
+                return Interop.Crypto.X509ChainNew(customTrust, Interop.Crypto.NewX509Stack());
+            }
+
+            return  Interop.Crypto.X509ChainNew(systemTrust, s_userRootStore.GetNativeCollection());
```
Nit: there's an extra space before Interop

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs:
```diff
@@ -69,21 +73,73 @@ internal sealed partial class ChainPal : IDisposable, IChainPal
+            if (trustMode == X509ChainTrustMode.CustomRootTrust)
+            {
+                CERT_CHAIN_ENGINE_CONFIG customChainEngine = new CERT_CHAIN_ENGINE_CONFIG();
+                customChainEngine.cbSize = Marshal.SizeOf<CERT_CHAIN_ENGINE_CONFIG>();
```
Please move the above two lines into the using block below.  Otherwise, we end up having this struct in scope outside the using refer to the handle that's disposed by the using.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs:
```diff
@@ -69,21 +73,73 @@ internal sealed partial class ChainPal : IDisposable, IChainPal
+        private static void AddDefaultCertificatesToStore(X509Certificate2Collection collectionStore, bool useMachineContext)
+        {
+            StoreLocation storeLocation = useMachineContext ? StoreLocation.LocalMachine: StoreLocation.CurrentUser;
+            foreach (StoreName storeName in (StoreName[]) Enum.GetValues(typeof(StoreName)))
```
This is expensive.  The array should be cached, and reflection should not be used to create it.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs:
```diff
@@ -69,21 +73,73 @@ internal sealed partial class ChainPal : IDisposable, IChainPal
+                    collectionStore.Add(certificate);
+                }
+
+                store.Close();
```
Can the `X509Store` be in a using block instead?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Interop.crypt32.cs:
```diff
@@ -237,13 +237,17 @@ public static unsafe byte[] EncodeObject(string lpszStructType, void* decoded)
+            CERT_CHAIN_ENGINE_CONFIG pConfig,
+            out IntPtr hChainEngine);
+
+        [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode, SetLastError = true)]
```
What's the purpose of SetLastError = true here?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Interop.crypt32.cs:
```diff
@@ -237,13 +237,17 @@ public static unsafe byte[] EncodeObject(string lpszStructType, void* decoded)
-        {
-            return CertGetCertificateChain((IntPtr)hChainEngine, pCertContext, pTime, hStore, ref pChainPara, dwFlags, pvReserved, out ppChainContext);
-        }
+        [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode, SetLastError = true)]
```
What's the purpose of SetLastError = true here?

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestChainTests.cs:
```diff
@@ -201,6 +218,85 @@ public static void ChainCertRequirements(bool useIntermed, bool? isCA, X509KeyUs
+                {
+                    chain.ChainPolicy.ExtraStore.Add(rootCert);
+                    chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                    chain.ChainPolicy.VerificationTime = DateTimeOffset.UtcNow.ToLocalTime().DateTime;
```
The VerificationTime value is already the Now value from when the Chain(Policy) was built.  The purpose of setting it explicitly in previous tests was to avoid clock skew, so here it should be set as something like `leafCert.NotBefore.AddSeconds(1)`

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestChainTests.cs:
```diff
@@ -10,27 +10,38 @@ namespace System.Security.Cryptography.X509Certificates.Tests.CertificateCreatio
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public static void CreateChain_ECC(bool useExtraStore)
```
The tests in this file are really just making sure that the CertificateRequest class is building certificates such that they properly chain.

The tests for using newly created chains to test functionality of X509Chain should really be in DynamicChainTests.cs .. it helps with scoping to see what component is failing.  (And once in DynamicChainTests.cs the tests can use `TestDataGenerator.MakeTestChain3` to make a 3-deep chain appear out of nowhere, or new helpers as warranted).

I think ideally the changes in this file would all be reverted, as it then maximizes how compatible this file is with .NET Framework.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestChainTests.cs:
```diff
@@ -201,6 +218,85 @@ public static void ChainCertRequirements(bool useIntermed, bool? isCA, X509KeyUs
         }
 
+        [Fact]
+        public static void BuildChainUntrustedRoot()
```
This test seems to be testing that adding the intermediate to the CustomTrustStore doesn't provide trust.  That's a good test, but the name doesn't really convey that.  `CustomRootTrustDoesNotTrustIntermediates` seems more appropriate (once this test is in DynamicChainTests)

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestChainTests.cs:
```diff
@@ -201,6 +218,85 @@ public static void ChainCertRequirements(bool useIntermed, bool? isCA, X509KeyUs
+                using (X509Chain chain = new X509Chain())
+                {
+                    chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                    chain.ChainPolicy.VerificationTime = new DateTime(2015, 10, 15, 12, 01, 01, DateTimeKind.Local);
```
Rather than hard-coding dates, please derive it from `cert.NotBefore` and/or `cert.NotAfter` of the cert that will be passed to chain.Build.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/CertificateRequestChainTests.cs:
```diff
@@ -201,6 +218,85 @@ public static void ChainCertRequirements(bool useIntermed, bool? isCA, X509KeyUs
+        }
+
+        [Fact]
+        public static void CustomStoreWithPublicCertificate()
```
The name on this one seems a little off, too.  `public static void SystemTrustedCertificateWithCustomRootTrust(bool trustCorrectRoot)` seems better.  If this one doesn't need any dynamic elements it can go straight into ChainTests.cs

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,88 @@ public static void BuildChainExtraStoreUntrustedRoot()
+                // Normal case
+                X509Chain chain = chainHolder.Chain;
+                chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                chain.ChainPolicy.VerificationTime = new DateTime(2015, 9, 22, 12, 25, 0);
```
Derive the time from the leaf certificate NotBefore/NotAfter.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,88 @@ public static void BuildChainExtraStoreUntrustedRoot()
+                X509Certificate2 intermediateCert = chain.ChainElements[1].Certificate;
+
+                //Trust mode set to custom, but no certificates added
+                chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
```
Ideally each sub-case here would be on a different chain object, and possibly from a different method (or the same method driven by theory inputs).

Currently this method is not disposing the certificates created from each prior chain, and if something fails it won't be obvious in the logs what went wrong (too many Assert.Falses, too many Assert.Equals(3), too many Assert.Equal(UntrustedRoot))

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,88 @@ public static void BuildChainExtraStoreUntrustedRoot()
+                chain.ChainPolicy.CustomTrustStore.Add(new X509Certificate2());
+                chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+
+                Assert.Throws<ArgumentException>("CustomTrustStore", () => chain.Build(testCert));
```
The two new throwing cases should be separate tests.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,25 +128,66 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
+            {
+                for (int i = 0; i < customTrustStore.Count; i++)
+                {
+                    AppleCertificatePal extraCertPal = (AppleCertificatePal)customTrustStore[i].Pal;
```
There should be a dynamic chain test where the intermediates and roots are added to CustomRootStore (and nothing to ExtraStore) to see if all platforms are building the chains the same way.

zwread commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:
```diff
@@ -125,6 +128,36 @@ public SafeX509ChainHandle SafeHandle
+                }
+
+                Interop.Crypto.X509StackAddMultiple(untrusted, systemTrust);
+                Interop.Crypto.X509StackAddMultiple(untrusted, s_userRootStore.GetNativeCollection());
```
Yes, we are adding them to the untrusted set so that we get an untrusted root rather than a partial chain. I will add a comment for this.

zwread commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/ChainPal.BuildChain.cs:
```diff
@@ -69,21 +73,73 @@ internal sealed partial class ChainPal : IDisposable, IChainPal
+
+                foreach (X509Certificate2 certificate in store.Certificates)
+                {
+                    collectionStore.Add(certificate);
```
I am adding the default store certificates to extra store. From MSDN 
> Exclusive trust mode allows applications to specify trust anchors and peer-trusted certificates for certificate chain validation. In the exclusive trust mode, the root store and the trusted people store on the system are ignored, and the anchors and certificates pointed to by the hExclusiveRoot and hExclusiveTrustedPeople members are used instead.

To avoid a partial chain I need to add the root store and trusted people store to the extraStore collection so that they can be passed in as untrusted. I will add a comment about this to prevent confusion.

bartonjs replied:
Ah, somehow I read that backwards (that you were adding to the stores, not draining the stores).

I'd rather that this didn't do this.  If Windows isn't looking at the default stores then we shouldn't either (and it would remove variations of this code from all three platforms).  I expect 99% of the use cases to be "if it isn't chained to this/one-of-these root(s) I don't want it", so PartialChain vs UntrustedRoot wouldn't matter much.

zwread replied:
Ok that makes things simpler. The one issue is that on osx I couldn't find a way not to check the default stores. Setting the trust anchor certificates only affects which certificates are trusted and not what stores are searched.

So Windows would produce a PartialChain and OSX UnstrustedRoot.

zachariahread committed:
Fix QA Issues

Fix formatting issues, revert changes to CertificateRequestChainTests,
Add tests to ChainTests and DynamicChainTests, add error handling,
dispose of objects, create SafeChainEngineHandle

Don't search through default system stores when
CustomTrustMode is set for linux and windows.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,142 @@ public static void BuildChainExtraStoreUntrustedRoot()
+
+                Assert.False(chain.Build(microsoftDotCom));
+                Assert.Equal(3, chain.ChainElements.Count);
+                Assert.Equal(X509ChainStatusFlags.UntrustedRoot, chain.AllStatusFlags());
```
Since this and the previous test only differ by the last two lines, please merge them (just branch the test internally using `if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))`)

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,142 @@ public static void BuildChainExtraStoreUntrustedRoot()
+
+            using (var microsoftDotCom = new X509Certificate2(TestData.MicrosoftDotComSslCertBytes))
+            using (var testCert = new X509Certificate2(Path.Combine("TestData", "test.pfx"), TestData.ChainPfxPassword))
+            using (var chainHolder = new ChainHolder())
```
Please use two different chainHolder/chain values.  One for the "find the intermediate and root" and the other for the test variation; rather than have all of these variations have potentially tainted values from the "find" invocation.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,142 @@ public static void BuildChainExtraStoreUntrustedRoot()
+        [Theory]
+        [InlineData(true)]
+        [InlineData(false)]
+        public static void BuildChainCustomTrustStoreWithExceptions(bool systemTrustMode)
```
This should just be two separate test methods, not a theory.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,142 @@ public static void BuildChainExtraStoreUntrustedRoot()
+        public static void BuildChainCustomTrustStoreWithExceptions(bool systemTrustMode)
+        {
+            using (var testCert = new X509Certificate2(Path.Combine("TestData", "test.pfx"), TestData.ChainPfxPassword))
+            using (ImportedCollection ic = Cert.Import(Path.Combine("TestData", "test.pfx"), TestData.ChainPfxPassword, X509KeyStorageFlags.DefaultKeySet))
```
This collection doesn't seem to be used at all..

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs:
```diff
@@ -270,6 +270,47 @@ public static void VerifyNumericStringSubject()
+                chain.ChainPolicy.VerificationTime = endEntityCert.NotBefore.AddSeconds(1);
+                chain.ChainPolicy.TrustMode = X509ChainTrustMode.CustomRootTrust;
+                chain.ChainPolicy.CustomTrustStore.Add(intermediateCert);
+                if(saveAllInCustomTrustStore)
```
Blank lines before control flow statements make me happy

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/DynamicChainTests.cs:
```diff
@@ -270,6 +270,47 @@ public static void VerifyNumericStringSubject()
+                Assert.Equal(3, chain.ChainElements.Count);
+                Assert.Equal(chainFlags, chain.AllStatusFlags());
+
+                chainHolder.DisposeChainElements();
```
This is done automatically by chainHolder being in a using statement.  It's only needed to be called manually if a second chain.Build is going to happen.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/tests/ChainTests.cs:
```diff
@@ -226,6 +226,142 @@ public static void BuildChainExtraStoreUntrustedRoot()
+                    case BuildChainCustomTrustStoreTestArguments.MultipleCalls:
+                        chain.ChainPolicy.CustomTrustStore.Add(intermediateCert);
+                        chain.ChainPolicy.CustomTrustStore.Add(rootCert);
+                        chain.Build(microsoftDotCom);
```
chainHolder.DisposeChainElements(); after this build (so that the objects it created are disposed before the next chain.Build)

zachariahread committed:
Fix SafeHandle classes and QA comments

zachariahread committed:
Fix SafeChainEngineHandle disposal

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/SafeHandles.cs:
```diff
@@ -238,4 +238,38 @@ protected sealed override bool ReleaseHandle()
+            SafeHandleCache<SafeChainEngineHandle>.GetInvalidHandle(() => new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_LOCAL_MACHINE));
+
+        public static SafeChainEngineHandle UserChainEngine =>
+            SafeHandleCache<SafeChainEngineHandle>.GetInvalidHandle(() => new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_CURRENT_USER));
```
These aren't modelling invalid handles, and only one of the two of them will be saved as "the" handle.

Instead they should just be

```C#
public static readonly SafeChainEngineHandle MachineChainEngine =
    new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_LOCAL_MACHINE);

public static readonly SafeChainEngineHandle UserChainEngine =
    new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_CURRENT_USER);
```

And Dispose:

```C#
protected override void Dispose(bool disposing)
{
    if (this != UserChainEngine && this != MachineChainEngine)
    {
        base.Dispose(disposing);
    }
}
```

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509Chain.cs:
```diff
@@ -112,6 +112,20 @@ internal bool Build(X509Certificate2 certificate, bool throwOnException)
+                    {
+                        if (customCertificate == null || customCertificate.Handle == IntPtr.Zero)
+                        {
+                            throw new ArgumentException(SR.Cryptography_InvalidTrustCertificate, nameof(_chainPolicy.CustomTrustStore));
```
This one should also be a CryptographicException, since it's not about an argument to this method.

bartonjs commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509ChainPolicy.cs:
```diff
@@ -68,6 +72,20 @@ public X509VerificationFlags VerificationFlags
+            set
+            {
+                if (value < X509ChainTrustMode.System || value > X509ChainTrustMode.CustomRootTrust)
+                    throw new ArgumentException(SR.Format(SR.Arg_EnumIllegalVal, nameof(value)));
```
I don't think there are tests for this exception.  Please add a test where it assigns `(X509ChainTrustMode)-1` and `(X509ChainTrustMode)2`.  (Probably ChainTests.cs, seems like a reasonable Theory (InlineData(-1), InlineData(2)))

zachariahread committed:
Fix linux and osx build issues and add test.

Add test for invalid X509ChainTrustMode and fix
SafeChainEngineHandle machine and user initilization.

zachariahread committed:
Fix build issues

zachariahread committed:
Set SafeChainEngineHandle to zero

stephentoub commented on src/System.Security.Cryptography.X509Certificates/ref/System.Security.Cryptography.X509Certificates.cs:
```diff
@@ -372,7 +374,7 @@ public partial struct X509ChainStatus
         private object _dummy;
         private int _dummyPrimitive;
-        public System.Security.Cryptography.X509Certificates.X509ChainStatusFlags Status { get { throw null; } set { } }
+        public System.Security.Cryptography.X509Certificates.X509ChainStatusFlags Status { readonly get { throw null; } set { } }
```
Was this API reviewed?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,60 +148,86 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
             AppleCertificatePal applePal = (AppleCertificatePal)cert;
-
-            safeHandles[0] = applePal.CertificateHandle;
+            safeHandles.Add(applePal.CertificateHandle);
```
Nit: does this need a local?  It could just be:
```C#
safeHandles.Add((AppleCertificatePal)cert).CertificateHandle);
```
or even:
```C#
var safeHandles = new List<SafeHandle> { (AppleCertificatePal)cert).CertificateHandle };
```

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,60 +148,86 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
                 for (int i = 0; i < extraStore.Count; i++)
                 {
                     AppleCertificatePal extraCertPal = (AppleCertificatePal)extraStore[i].Pal;
+                    safeHandles.Add(extraCertPal.CertificateHandle);
```
Same locals question

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs:
```diff
@@ -120,60 +148,86 @@ private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
+                {
+                    // Only adds non self issued certs to the untrusted certs array. Trusted self signed
+                    // certs will be added to the custom certs array.
+                    if (!cert.SubjectName.RawData.ContentsEqual(cert.IssuerName.RawData))
```
Why is this `if` inside the `for` loop rather than outside of it?

bartonjs replied:
Hm. I think it's supposed to be using `customTrustStore[i]` instead of cert.  Good eye.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs:
```diff
@@ -125,6 +128,32 @@ public SafeX509ChainHandle SafeHandle
+                    x509Chain = Interop.Crypto.X509ChainNew(customTrust, SafeX509StackHandle.InvalidHandle);
+                }
+
+                return x509Chain;
```
Why define and return this outside of the using, rather than just doing `return Interop.Crypto.X509ChainNew(...);` inside the using?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Primitives.cs:
```diff
@@ -768,4 +768,33 @@ internal enum CryptAcquireFlags : int
+        public int CycleDetectionModulus;
+        public IntPtr hExclusiveRoot;
+        public IntPtr hExclusiveTrustedPeople;
+    }
```
This doesn't match the definition at https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-cert_chain_engine_config, which has an extra DWORD dwExclusiveFlags field at the end.

zwread replied:
The dwExclusiveFlags field is only supported by windows 8 and on. I can add 2 structs and check the os when deciding which to use.

bartonjs replied:
Since Windows 8+ understand the sizeof/versioning from Win7, I'd leave the struct as-is.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/SafeHandles.cs:
```diff
@@ -238,4 +238,39 @@ protected sealed override bool ReleaseHandle()
+            new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_LOCAL_MACHINE);
+
+        public static SafeChainEngineHandle UserChainEngine =>
+            new SafeChainEngineHandle((IntPtr)ChainEngine.HCCE_CURRENT_USER);
```
I'm confused by these.  Did you mean to make these properties returning new instances on each call?  If so, the Dispose implementation is bad, since it a) compares this instance against new safe handle instances that it'll never equal, and b) allocates these two handles without disposing of them.  Which leads me to believe you intended these to be `=` rather than `=>`, in which case they should also be made `readonly`.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509ChainPolicy.cs:
```diff
@@ -26,6 +28,8 @@ public X509ChainPolicy()
 
         public X509Certificate2Collection ExtraStore => _extraStore ?? (_extraStore = new X509Certificate2Collection());
 
+        public X509Certificate2Collection CustomTrustStore => _customTrustStore ?? (_customTrustStore = new X509Certificate2Collection());
```
```suggestion
        public X509Certificate2Collection CustomTrustStore => _customTrustStore ??= new X509Certificate2Collection();
```

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/X509ChainTrustMode.cs:
```diff
@@ -0,0 +1,15 @@
+        System = 0,
+        CustomRootTrust = 1,
+    }
+}
```
Nit: missing newline at end of file

zachariahread committed:
Addressed QA comments

Remove unnecessary local variables
Fix checking if self signed certificate in ChainPal
Fix default SafeChainEngineHandles

zachariahread committed:
Check for empty customtrustroot windows

zachariahread committed:
Add test with no certificates when in CustomRootTrust

zachariahread committed:
QA comments

Remove unsafe from crypt32
Correctly dispose of certificates in tests

stephentoub commented on src/System.Net.Security/src/System.Net.Security.csproj:
```diff
@@ -387,6 +387,9 @@
     </Compile>
+    <Compile Include="$(CommonPath)\Microsoft\Win32\SafeHandles\SafeHandleCache.cs">
+      <Link>Common\Microsoft\Win32\SafeHandles\SafeHandleCache.cs</Link>
+    </Compile>
```
Why are these additions to System.Net.Security and System.Net.Http necessary?

zwread replied:
To use SafeHandlesCache in Interop.X509Stack.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Interop.crypt32.cs:
```diff
@@ -237,13 +237,25 @@ public static unsafe byte[] EncodeObject(string lpszStructType, void* decoded)
-        private static extern unsafe bool CertGetCertificateChain(IntPtr hChainEngine, SafeCertContextHandle pCertContext, FILETIME* pTime, SafeCertStoreHandle hStore, [In] ref CERT_CHAIN_PARA pChainPara, CertChainFlags dwFlags, IntPtr pvReserved, out SafeX509ChainHandle ppChainContext);
+        private static extern bool CertCreateCertificateChainEngine(ref CERT_CHAIN_ENGINE_CONFIG pConfig, out SafeChainEngineHandle hChainEngineHandle);
+
+        [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode)]
```
This CharSet isn't necessary, is it?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Windows/Native/Interop.crypt32.cs:
```diff
@@ -237,13 +237,25 @@ public static unsafe byte[] EncodeObject(string lpszStructType, void* decoded)
+        [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode)]
+        public static extern void CertFreeCertificateChainEngine(IntPtr hChainEngine);
+
+        [DllImport(Libraries.Crypt32, CharSet = CharSet.Unicode, SetLastError = true)]
```
Is this CharSet necessary?

zachariahread committed:
Remove CharSet from library reference

bartonjs commented:
Thanks for slogging through this, @zwread!

zwread commented:
Thank you for all the help and feedback @bartonjs @stephentoub
